GLRenderer/GLRenderer/Textures/Texture.cpp

```diff
@@ -122,77 +140,101 @@ void C_Texture::SetWrap(Renderer::E_WrapFunction wrapS, Renderer::E_WrapFunction
 	unbind();
 }
 
 //=================================================================================
 void C_Texture::SetTexParameter(GLenum pname, GLint value)
 {
 	bind();
-	glTexParameteri(m_target, pname, value);
+	if (!m_bIsTexture)
+		glTexParameteri(GetTextureType(m_Desc.type), pname, value);
+	else
+		glTextureParameteri(m_texture, pname, value);
 	unbind();
 }
 
 //=================================================================================
 void C_Texture::GenerateMipMaps()
 {
 	bind();
-	glGenerateMipmap(m_target);
-	unbind();
-}
-
-//=================================================================================
-void C_Texture::SetTexData2D(int level, const Renderer::MeshData::Texture& tex)
-{
-	bind();
-	SetDimensions({tex.width, tex.height});
-	static_assert(std::is_same_v<std::uint8_t, decltype(tex.data)::element_type>, "Format have ben changed.");
-	m_Format = Renderer::E_TextureFormat::RGBA8i;
-
-	glTexImage2D(m_target, level,
-				 GetOpenGLInternalFormat(m_Format),		  // internal format
-				 (GLsizei)tex.width, (GLsizei)tex.height, // dimensions
-				 0,										  // border
-				 GL_RGBA,								  // format
-				 T_TypeToGL<decltype(tex.data)::element_type>::value,
-				 tex.data.get() // data
-	);
+	if (!m_bIsTexture)
+		glGenerateMipmap(GetTextureType(m_Desc.type));
+	else
+		glGenerateTextureMipmap(m_texture);
 	unbind();
 }
 
 //=================================================================================
 void C_Texture::SetTexData2D(int level, const Renderer::I_TextureViewStorage* tex)
 {
-	bind();
-	SetDimensions(tex->GetDimensions());
-
-	glTexImage2D(m_target, level,
-				 GetOpenGLInternalFormat(m_Format), // internal format
-				 tex->GetDimensions().x, tex->GetDimensions().y,
-				 0,								// border
-				 GetFormat(tex->GetChannels()), // format
-				 GetUnderlyingType(tex),		// TODO
-				 tex->GetData());				// data
-	unbind();
+	GLE_ASSERT(tex, "This should be smth like reference");
+	if (!m_bIsTexture)
+	{
+		bind();
+		SetDimensions(tex->GetDimensions());
+
+		glTexImage2D(GetTextureType(m_Desc.type), level,
+					 GetOpenGLInternalFormat(m_Desc.format), // internal format
+					 tex->GetDimensions().x, tex->GetDimensions().y,
+					 0,								// border
+					 GetFormat(tex->GetChannels(), Renderer::IsIntegral(tex->GetStorageType())), // format
+					 GetUnderlyingType(tex),		// TODO
+					 tex->GetData());															 // data
+		unbind();
+	}
+	else
+	{
+		// https://www.khronos.org/registry/OpenGL/specs/gl/glspec45.compatibility.pdf Table 8.35
+		glTextureSubImage2D(m_texture,
+							0,												//!< Level
+							0, 0,											//!< Offsets
+							tex->GetDimensions().x, tex->GetDimensions().y, //!< Dimmensions
+							GetFormat(tex->GetChannels(), Renderer::IsIntegral(tex->GetStorageType())),
+							GetUnderlyingType(tex),
+							tex->GetData());
+	}
+	// automatic mip-maps generation
+	if (m_Desc.m_Levels > 1)
+		glGenerateTextureMipmap(m_texture);
```

Comment:
Really nice feature, but if I allow users to set each level separately I should not overwrite it by mip map generation.